### PR TITLE
Add a timeout to the wait for transforms step of the costmap activation.

### PR DESCRIPTION
**Pull request by kfabian**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3749 |
| Primary OS tested on | Ubuntu 22.04 |
| Robotic platform tested on | only tested without robot by optionally manually providing the transform |

---

## Description of contribution in a few bullet points

* Previously the `on_activate` method of the costmap was blocking indefinitely if the transform from the robot base frame to the global frame is not available
* This adds a timeout so that the transition from `configured` to `active` fails if the transform is not available
* The timeout can be configured by a costmap parameter
* The controller server transition from `configured` to `active` fails if the corresponding costmap transitions fails

## Description of documentation updates required from your changes

* There is a new costmap parameter: `wait_for_transforms_timeout`
---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_controller/src/controller_server.cpp
Comment (SteveMacenski) [inline nav2_controller/src/controller_server.cpp:245]: For planner server too?

File: nav2_costmap_2d/src/costmap_2d_ros.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:132]: How about `initial_transform_timeout`?

File: nav2_costmap_2d/src/costmap_2d_ros.cpp
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:301]: Don't deactivate. This will trigger the error state and its not the role of a failed activation to deactivate things according to the ROS 2 lifecycle state machine docs

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally speaking, LGTM!

File: nav2_costmap_2d/src/costmap_2d_ros.cpp
Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:132]: ^ also needs a docs PR for navigation.ros.org adding it + a migration guide entry about the change. Please submit that for the next PR review cycle (I'll probably be able to merge the pair together)

**Comment by SteveMacenski** [comment]
Alot of the system tests failed. I'm going to rerun them again in case that was a fluke, but that might be pointing to some activation errors this PR introduces

File: nav2_controller/src/controller_server.cpp
Comment (kfabian) [reply nav2_controller/src/controller_server.cpp:245]: I haven't considered the planner server, it should get the same change.

File: nav2_costmap_2d/src/costmap_2d_ros.cpp
Comment (kfabian) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:301]: I am deactivating the lifecycle publishers not lifecycle nodes, I don't think this will trigger the error state. According to the lifecycle design doc, a failure in onActivate will result in the inactive state. So I think if I won't deactivate the publishers they will still be active despite the node being in the inactive state. It would be easiest though to just change the order inside the `on_activate` method, so that the transforms are checked at the top of the method.

**Comment by kfabian** [comment]
> Alot of the system tests failed. I'm going to rerun them again in case that was a fluke, but that might be pointing to some activation errors this PR introduces

Maybe the new timeout is just too low for tests with the gazebo simulation? Unfortunately I could not get the turtlebot simulation running on my Ubuntu arm64 virtual machine.

**Review by SteveMacenski — APPROVED** [review]
Still needs docs PR I described but otherwise LGTM!

**Comment by kfabian** [comment]
@SteveMacenski the CI is complaining about reduced test coverage. Looking at the costmap test folder I'm wondering where a suitable test should be located.

**Comment by SteveMacenski** [comment]
A new file would be fine!

**Comment by kfabian** [comment]
> Still needs docs PR I described but otherwise LGTM!

I created the docs PR: https://github.com/ros-planning/navigation.ros.org/pull/477

**Comment by SteveMacenski** [comment]
Waiting on CI

**Comment by SteveMacenski** [comment]
Thanks!!

**Pull request merged**
